### PR TITLE
[SshV0] Migrated to ssh2-sftp-client

### PR DESCRIPTION
**Pull request by sdobrodeev**
**Task name**: SshV0

**Description**: Migrated from obsolete scp2 library

**Documentation changes required:** (Y/N) N

**Added unit tests:** (Y/N) N

**Attached related issue:** (Y/N) <Please add link to related issue here>

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

### CODE REVIEW

File: Tasks/SshV0/task.loc.json [hunk 1]
```diff
@@ -18,7 +18,7 @@
   "version": {
     "Major": 0,
     "Minor": 176,
-    "Patch": 0
+    "Patch": 1
   },
   "demands": [],
   "minimumAgentVersion": "2.102.0",
```
Comment (anatolybolshakov) [inline Tasks/SshV0/task.loc.json:24]: Should we increase minimumAgentVersion also?

File: Tasks/SshV0/ssh2helpers.ts [hunk 2]
```diff
@@ -63,23 +63,33 @@ function handleStreamClose(command: string, stdErrWritten: boolean, defer: Q.Def
...
+ * Uses sftp to copy a file to remote machine
+ * @param src
+ * @param dest
+ * @param sftpConfig
  * @returns {Promise<string>}
  */
-export function copyScriptToRemoteMachine(scriptFile: string, scpConfig: any): Q.Promise<string> {
+export async function copyScriptToRemoteMachine(src: string, dest: string, sftpConfig: any): Promise<string> {
     const defer = Q.defer<string>();
+    const sftpClient = new SftpClient();;
```
Comment (anatolybolshakov) [inline Tasks/SshV0/ssh2helpers.ts:74]: Nit: doubled semicolon

File: Tasks/SshV0/ssh2helpers.ts [hunk 3]
```diff
@@ -63,23 +63,33 @@ function handleStreamClose(command: string, stdErrWritten: boolean, defer: Q.Def
...
- * @param scriptFile
- * @param scpConfig
+ * Uses sftp to copy a file to remote machine
+ * @param src
+ * @param dest
+ * @param sftpConfig
  * @returns {Promise<string>}
  */
-export function copyScriptToRemoteMachine(scriptFile: string, scpConfig: any): Q.Promise<string> {
+export async function copyScriptToRemoteMachine(src: string, dest: string, sftpConfig: any): Promise<string> {
```
Comment (anatolybolshakov) [inline Tasks/SshV0/ssh2helpers.ts:72]: Could you please define some interface for sftpConfig? Or maybe there some types we could import, could we use https://www.npmjs.com/package/@types/ssh2-sftp-client?

**Review by alexander-prozorov — APPROVED** [review]
LGTM, thanks

File: Tasks/SshV0/ssh2helpers.ts [hunk 4]
```diff
@@ -63,23 +63,33 @@ function handleStreamClose(command: string, stdErrWritten: boolean, defer: Q.Def
...
-        } else {
-            tl.debug('Copied script file to remote machine at: ' + scpConfig.path);
-            defer.resolve(scpConfig.path);
-        }
-    });
-
+    try {
+        sftpClient.on('error', (err) => {
+            tl.debug('sftpClient: Ignoring error diconnecting: ' + err);
+        }); // ignore logout errors; see: https://github.com/mscdex/node-imap/issues/695
```
Comment (alexander-smolyakov) [inline Tasks/SshV0/ssh2helpers.ts:88]: Nit: I assume it will be better to provide more details in the code comments instead of a link to the GitHub issue.

**Review by alexander-smolyakov — COMMENTED** [review]
Please take a look at the comments, other changes LGTM.

File: Tasks/SshV0/ssh2helpers.ts [hunk 5]
```diff
@@ -63,23 +63,33 @@ function handleStreamClose(command: string, stdErrWritten: boolean, defer: Q.Def
...
-            defer.reject(tl.loc('RemoteCopyFailed', err));
-        } else {
-            tl.debug('Copied script file to remote machine at: ' + scpConfig.path);
-            defer.resolve(scpConfig.path);
-        }
-    });
-
+    try {
+        sftpClient.on('error', (err) => {
+            tl.debug('sftpClient: Ignoring error diconnecting: ' + err);
```
[the remark this answers is not in the document]
Comment (anatolybolshakov) [reply Tasks/SshV0/ssh2helpers.ts]: Thanks! Fixed

File: Tasks/SshV0/ssh2helpers.ts (hunk 4 above)
Comment (anatolybolshakov) [reply Tasks/SshV0/ssh2helpers.ts]: Added some more description here

**Review by alexander-smolyakov — APPROVED** [review]
LGTM, thanks!

**Comment by anatolybolshakov** [comment]
Tested with canary and other cases, going to add canary tests for SSH task

**Review by anatolybolshakov — APPROVED** [review]

**Pull request merged**